drivers/staging/vc04_services/bcm2835-isp/bcm2835-v4l2-isp.c

```diff
@@ -0,0 +1,1808 @@
+
+	if (!NODE_IS_OUTPUT(node))
+		return -EINVAL;
+
+	return vidioc_enum_fmt(file, priv, f);
+}
+
+static int bcm2835_isp_node_try_fmt_vid_cap(struct file *file, void *priv,
+					    struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (!NODE_IS_CAPTURE(node))
+		return -EINVAL;
+
+	return vidioc_try_fmt(node, f, fmt);
+}
+
+static int bcm2835_isp_node_try_fmt_meta_cap(struct file *file, void *priv,
+					     struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (!NODE_IS_STATS(node))
+		return -EINVAL;
+
+	fmt = find_format(f, node);
+	if (!fmt) {
+		f->fmt.meta.dataformat = V4L2_META_FMT_BCM2835_ISP_STATS;
+		fmt = find_format(f, node);
+	}
+	f->fmt.meta.buffersize = get_port_data(node)->minimum_buffer.size;
+
+	return 0;
+}
+
+static int bcm2835_isp_node_try_fmt_vid_out(struct file *file, void *priv,
+					    struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_fmt *fmt;
+
+	if (!NODE_IS_OUTPUT(node))
+		return -EINVAL;
+
+	return vidioc_try_fmt(node, f, fmt);
+}
+
+static int bcm2835_isp_node_s_fmt_vid_cap(struct file *file, void *priv,
+					  struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	ret = bcm2835_isp_node_try_fmt_vid_cap(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set capture format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
+
+static int bcm2835_isp_node_s_fmt_meta_cap(struct file *file, void *priv,
+					   struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	ret = bcm2835_isp_node_try_fmt_meta_cap(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set meta format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
+
+static int bcm2835_isp_node_s_fmt_vid_out(struct file *file, void *priv,
+					  struct v4l2_format *f)
+{
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	int ret;
+
+	ret = bcm2835_isp_node_try_fmt_vid_out(file, priv, f);
+	if (ret)
+		return ret;
+
+	v4l2_dbg(1, debug, &node_get_dev(node)->v4l2_dev,
+		 "%s: Set output format for node %s[%d]\n",
+		 __func__, node->name, node->id);
+
+	return populate_qdata_fmt(f, node);
+}
+
+static int bcm2835_isp_node_s_selection(struct file *file, void *fh,
+					struct v4l2_selection *s)
+{
+	struct mmal_parameter_crop_t crop;
+	struct bcm2835_isp_node *node = video_drvdata(file);
+	struct bcm2835_isp_dev *dev = node_get_dev(node);
+	struct vchiq_mmal_port *port = get_port_data(node);
+
+	/* This return value is required fro V4L2 compliance. */
+	if (NODE_IS_STATS(node))
+		return -ENOTTY;
+
+	if (!NODE_IS_OUTPUT(node) || s->type != V4L2_BUF_TYPE_VIDEO_OUTPUT)
```

Comment:
The first case should be done by disabling the ioctl.
The second case needs to remain.